### PR TITLE
Call Caching v3

### PR DESCRIPTION
**Pull request by Horneth**

### CODE REVIEW

File: engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/ExecutionStore.scala [hunk 1]
```diff
@@ -80,7 +82,14 @@ final case class ExecutionStore(private val statusStore: Map[JobKey, ExecutionSt
...
   }
 
-  def runnableScopes = keysWithStatus(NotStarted) filter arePrerequisitesDone
+  /**
+    * Returns the list of jobs ready to be run, along with a Boolean indicating whether or not the list has been truncated.
+    * The size of the list will be MaxJobsToStartPerTick at most. If more jobs where found runnable, the boolean will be true, otherwise false.
+    */
+  def runnableScopes: (List[JobKey], Boolean) = {
+    val readyToStart = keysWithStatus(NotStarted).toStream filter arePrerequisitesDone
+    (readyToStart.take(ExecutionStore.MaxJobsToStartPerTick).toList, readyToStart.size > ExecutionStore.MaxJobsToStartPerTick)
```
Comment (Horneth) [inline engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/ExecutionStore.scala]: Note to self: fix this: readyToStart.size evaluates the whole stream which negates the reason of making it a stream entirely.

**Comment by cjllanwarne** [comment]
@Horneth could you give a quick big-picture synopsis of what this PR is?

**Comment by Horneth** [comment]
@cjllanwarne Yep, it would be easier to do on a white board I was thinking of tech talking it maybe if that's ok ?

File: engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/callcaching/CallCacheHashingJobActor.scala [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+    * Aggregates hash results together in a predictable and reproducible manner.
+    * Purposefully takes an Iterable, because it will be sorted appropriately in this method
+    * to ensure deterministic result, so the type of collection doesn't matter.
+    * Only aggregates hash values for which checkForHitOrMiss is true.
+    * If several hash keys are identical, the result of this method is undefined.
+    */
+  private def calculateHashAggregation(hashes: Iterable[HashResult], messageDigest: MessageDigest) = {
+    val sortedHashes = hashes.toList.filter(_.hashKey.checkForHitOrMiss).sortBy(_.hashKey.key).map(_.hashValue.value.getBytes)
+    sortedHashes foreach messageDigest.update
+    DatatypeConverter.printHexBinary(messageDigest.digest())
```
Comment (mcovarr) [inline engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/callcaching/CallCacheHashingJobActor.scala]: This is consistent with the SQL migration?

File: engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/callcaching/CallCacheHashingJobActor.scala [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+/**
+  * Actor responsible for calculating individual as well as aggregated hashes for a job.
+  * First calculate the initial hashes (individual and aggregated), and send them to its parent
+  * as well as the callCacheReadingJobActor if one was provided.
+  * From there, either wait for a NextBatchOfFileHashesRequest message from the callCacheReadingJobActor to trigger the next
+  * batch of file hash requests, or send it itself if there is no callCacheReadingJobActor.
+  * When we get all the file hashes for one batch, send those hashes to the CCRead actor and wait for the next NextBatchOfFileHashesRequest.
+  * If it was the last batch and all the files have been hashed, send all the hashes along with the aggregated file hash.
+  * If at any point the callCacheReadingJobActor dies (either because it decided this was a CacheMiss or it found a successful
+  * CacheHit), either keep hashing the files if writeToCache is true, or die if it's not.
```
Comment (mcovarr) [inline engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/callcaching/CallCacheHashingJobActor.scala]: so CCRJA is communicating with a different actor (EJEA?) for hit / miss status?

File: engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/callcaching/CallCacheHashingJobActor.scala [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+  }
+
+  final case class CallCacheHashingJobActorData(
+                                           fileHashRequestsRemaining: List[List[SingleFileHashRequest]],
+                                           fileHashResults: List[HashResult],
+                                           callCacheReadingJobActor: Option[ActorRef]
+                                         ) {
+    private val md5Digest = MessageDigest.getInstance("MD5")
+
+    def withFileHash(hashResult: HashResult): (CallCacheHashingJobActorData, Option[CCHJAFileHashResponse]) = {
```
Comment (mcovarr) [inline engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/callcaching/CallCacheHashingJobActor.scala]: type?

File: engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/callcaching/CallCacheReadingJobActor.scala [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+      callCacheReadActor ! CacheLookupRequest(AggregatedCallHashes(data.initialHash, aggregatedFileHash), data.currentHitNumber)
+      goto(WaitingForCacheHitOrMiss) using data.withFileHash(aggregatedFileHash)
+    case Event(NoFileHashesResult, data: CCRJAWithData) =>
+      callCacheReadActor ! CacheLookupRequest(AggregatedCallHashes(data.initialHash, None), data.currentHitNumber)
+      goto(WaitingForCacheHitOrMiss)
+  }
+  
+  when(WaitingForCacheHitOrMiss) {
+    case Event(CacheLookupNextHit(hit), data: CCRJAWithData) =>
+      context.parent ! CacheHit(hit)
```
Comment (mcovarr) [inline engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/callcaching/CallCacheReadingJobActor.scala]: Ok, this answers my question about the CCRJA communicating with its parent.  😄  Wouldn't hurt to drop a comment back in the CCHJA, this is all kind of a lot to take in.

**Comment by mcovarr** [comment]
:+1: as I mentioned in standup, I'm happy to rebase this if #2244 merges first.

File: CHANGELOG.md [hunk 6]
```diff
@@ -2,6 +2,25 @@
...
 ## 27
 
+### Migration
+
+* Call Caching has been improved in this version of Cromwell, specifically the time needed to determine whether or not a job can be cached
+ has drastically decreased. To achieve that the database schema has been modified and a migration is required in order to preserve the pre-existing cached jobs.
+ This migration is relatively fast compared to previous migrations. To get an idea of the time needed, look at the size of your `CALL_CACHING_HASH_ENTRY` table.
+ As a landmark, it takes 1 minute for a table with 6 million rows.
+ The migration will only be executed on MySQL. Other databases will lose their previous cached jobs.
+ In order to run properly on MySQL, the following flag needs to be adjusted: https://dev.mysql.com/doc/refman/5.5/en/server-system-variables.html#sysvar_group_concat_max_len
```
Comment (Horneth) [inline CHANGELOG.md]: The SQL query fails so the migration fails

Comment (Horneth) [inline CHANGELOG.md]: Provided that you have a call for which the concatenated hash values is greater than the default value for the flag which is 1024

File: engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/callcaching/CallCacheHashingJobActor.scala [hunk 7]
```diff
@@ -0,0 +1,267 @@
...
+                               writeToCache: Boolean
+                              ) extends LoggingFSM[CallCacheHashingJobActorState, CallCacheHashingJobActorData] {
+
+  val fileHashingActor = makeFileHashingActor()
+
+  // Watch the read actor, as it will die when it's done (cache miss or successful cache hit)
+  // When that happens we want to either stop if writeToCache is false, or keep going
+  callCacheReadingJobActor map context.watch
+
+  initializeCCHJA()
```
Comment (Horneth) [inline engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/callcaching/CallCacheHashingJobActor.scala]: yeah that's what I thought too but `initialize` is an `FSM` public method so akka isn't happy if you create your own.

File: engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/callcaching/CallCacheHashingJobActor.scala (hunk 2 above)
Comment (Horneth) [inline engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/callcaching/CallCacheHashingJobActor.scala]: Yes at least to the extent of the testing I've done they yield the same.
I ran this on the 10K mock wdl database after the migration and it call cached immediately.
`sortedHashes foreach messageDigest.update` is equivalent to md5ing the concatenation.
`DatatypeConverter.printHexBinary` just prints the bytes in hexadecimal string format.

File: CHANGELOG.md (hunk 6 above)
Comment (mcovarr) [inline CHANGELOG.md]: How long does this query take to run?  Is it something we could do in a precondition with a custom `onFailMessage` in the off chance that the user hasn't carefully read the CHANGELOG.md before migrating?  😄

Comment (Horneth) [inline CHANGELOG.md]: Oh yeah that's a good idea ! There might also be another way to do this that doesn't involve GROUP_CONCAT and doesn't need to adjust the flag but my SQL is not powerful enough..

File: engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/callcaching/CallCacheHashingJobActor.scala (hunk 7 above)
Comment (mcovarr) [inline engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/callcaching/CallCacheHashingJobActor.scala]: 😢

File: engine/src/test/scala/cromwell/engine/workflow/lifecycle/execution/callcaching/CallCacheHashingJobActorSpec.scala [hunk 8]
```diff
@@ -0,0 +1,292 @@
...
+
+    fileHashingActor.expectMsg(fileHashRequest1)
+    fileHashingActor.expectMsg(fileHashRequest2)
+    cchja.stateName shouldBe HashingFiles
+  }
+
+  it should "send NoFileHashesResult and stop if there are no input files" in {
+    val parent = TestProbe()
+    val callCacheReadProbe = TestProbe()
+    // self acts as the parent
```
Comment (Horneth) [inline engine/src/test/scala/cromwell/engine/workflow/lifecycle/execution/callcaching/CallCacheHashingJobActorSpec.scala]: Note to self: remove those comments they don't apply anymore as we have explicit parent `TestProbe`s now

File: engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/callcaching/CallCacheHashingJobActor.scala (hunk 4 above)
Comment (Horneth) [inline engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/callcaching/CallCacheHashingJobActor.scala]: It seems a bit superfluous for this one as this method is only ever used once inside this actor, and hence its result parsed once ? Also the name would be `NewDataAndMaybeResponse` :/
What about a scaladoc that says what the pair corresponds to ?

**Comment by mcovarr** [comment]
Rebased and a minor migration fix added.  I also removed the non-default failure and error handling on the migration but did not add in a special error message.

**Comment by Horneth** [comment]
@mcovarr thanks for rebasing !

File: engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/ExecutionStore.scala [hunk 9]
```diff
@@ -80,7 +84,18 @@ final case class ExecutionStore(private val statusStore: Map[JobKey, ExecutionSt
...
+    * Returns the list of jobs ready to be run, along with a Boolean indicating whether or not the list has been truncated.
+    * The size of the list will be MaxJobsToStartPerTick at most. If more jobs where found runnable, the boolean will be true, otherwise false.
+    */
+  def runnableScopes: RunnableScopes = {
+    val readyToStart = keysWithStatus(NotStarted).toStream filter arePrerequisitesDone
+    // Compute the first ExecutionStore.MaxJobsToStartPerTick + 1 runnable scopes
+    val scopesToStartPlusOne = readyToStart.take(ExecutionStore.MaxJobsToStartPerTick + 1).toList
+    // Only take the first ExecutionStore.MaxJobsToStartPerTick from the above list.
+    // Use the fact that we took one more to determine whether or not we truncated the result.
+    RunnableScopes(scopesToStartPlusOne.take(ExecutionStore.MaxJobsToStartPerTick), scopesToStartPlusOne.size > ExecutionStore.MaxJobsToStartPerTick)
```
Comment (cjllanwarne) [inline engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/ExecutionStore.scala]: I don't get this indirection with `scopesToStartPlusOne`. Why not just do:
```
RunnableScopes(readyToStart.take(ExecutionStore.MaxJobsToStartPerTick), 
     readyToStart.size > ExecutionStore.MaxJobsToStartPerTick)
```

File: engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/callcaching/CallCache.scala [hunk 10]
```diff
@@ -62,20 +69,31 @@ class CallCache(database: CallCachingSqlDatabase) {
...
-  def callCachingEntryIdsMatchingHashes(callCacheHashes: CallCacheHashes)(implicit ec: ExecutionContext): Future[Set[CallCachingEntryId]] = {
-    callCachingEntryIdsMatchingHashes(NonEmptyList.fromListUnsafe(callCacheHashes.hashes.toList))
+  def hasBaseAggregatedHashMatch(baseAggregatedHash: String)(implicit ec: ExecutionContext): Future[Boolean] = {
+    database.hasMatchingCallCachingEntriesForBaseAggregation(baseAggregatedHash)
+  }
+  
+  def hasKeyValuePairHashMatch(hashes: Set[HashResult])(implicit ec: ExecutionContext): Future[Boolean] = {
+    hashes.toList match {
+        // Empty hashes should match nothing
+      case Nil => Future.successful(false)
```
Comment (cjllanwarne) [inline engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/callcaching/CallCache.scala]: Really? If a task has no input files then we want it to match anything else with no input files

File: engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/callcaching/CallCacheHashingJobActor.scala (hunk 7 above)
Comment (cjllanwarne) [inline engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/callcaching/CallCacheHashingJobActor.scala]: British English to the rescue? `initialise` 🇬🇧

File: engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/callcaching/CallCacheHashingJobActor.scala [hunk 11]
```diff
@@ -0,0 +1,273 @@
...
+
+  override def preStart(): Unit = {
+    if (callCacheReadingJobActor.isEmpty && !writeToCache) {
+      log.error("Programmer error ! There is no reason to have a hashing actor if both read and write to cache are off")
+      context stop self
+    }
+    super.preStart()
```
Comment (cjllanwarne) [inline engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/callcaching/CallCacheHashingJobActor.scala]: Is this safe after a `context stop self`?

File: engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/callcaching/CallCacheHashingJobActor.scala [hunk 12]
```diff
@@ -0,0 +1,273 @@
...
+      stopAndStay(None)
+    case Event(Terminated(actor), data) if writeToCache =>
+      if (stateName == WaitingForHashFileRequest) self ! NextBatchOfFileHashesRequest
+      stay() using data.copy(callCacheReadingJobActor = None)
+    case Event(error: HashingFailedMessage, data) =>
+      log.error(error.reason, s"Failed to hash ${error.file}")
+      sendToCCReadActor(error, data)
+      context.parent ! error
+      stopAndStay(None)
+  }
```
Comment (cjllanwarne) [inline engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/callcaching/CallCacheHashingJobActor.scala]: Any value in a catch-all `case Event(other, data) => log.error("Panic! $other $data")`

File: engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/callcaching/CallCacheHashingJobActor.scala [hunk 13]
```diff
@@ -0,0 +1,273 @@
...
+        case (newData, None) =>
+          stay() using newData
+      }
+  }
+
+  whenUnhandled {
+    case Event(Terminated(actor), data) if !writeToCache =>
+      stopAndStay(None)
+    case Event(Terminated(actor), data) if writeToCache =>
+      if (stateName == WaitingForHashFileRequest) self ! NextBatchOfFileHashesRequest
```
Comment (cjllanwarne) [inline engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/callcaching/CallCacheHashingJobActor.scala]: Hmm, would this be better handled in the `when(stateName)` methods, rather than with a `stateName`-branch in the `whenUnhandled`?

File: engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/callcaching/CallCacheHashingJobActor.scala [hunk 14]
```diff
@@ -0,0 +1,273 @@
...
+
+  /**
+    * Aggregates hash results together in a predictable and reproducible manner.
+    * Purposefully takes an Iterable, because it will be sorted appropriately in this method
+    * to ensure deterministic result, so the type of collection doesn't matter.
+    * Only aggregates hash values for which checkForHitOrMiss is true.
+    * If several hash keys are identical, the result of this method is undefined.
+    */
+  private def calculateHashAggregation(hashes: Iterable[HashResult], messageDigest: MessageDigest) = {
+    val sortedHashes = hashes.toList.filter(_.hashKey.checkForHitOrMiss).sortBy(_.hashKey.key).map(_.hashValue.value.getBytes)
```
Comment (cjllanwarne) [inline engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/callcaching/CallCacheHashingJobActor.scala]: As above, I think it's important to make sure the keys are part of the hash, as well as the values

Comment (cjllanwarne) [inline engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/callcaching/CallCacheHashingJobActor.scala]: I assume that `checkForHitOrMiss` is the same as the set of runtime attributes that you specified in the migration query?

File: engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/callcaching/EngineJobHashingActor.scala [hunk 15]
```diff
@@ -1,282 +1,126 @@
...
+
+  override def receive = {
+    case initialHashResult: InitialHashingResult => initialHash = Option(initialHashResult)
+    case finalFileHashResult: FinalFileHashingResult => sendHashes(finalFileHashResult)
+    case CacheMiss => receiver ! CacheMiss
+    case hit: CacheHit => receiver ! hit
+    case NextHit =>
+      callCacheReadingJobActor match {
+        case Some(readActor) =>
+          readActor ! NextHit
```
Comment (cjllanwarne) [inline engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/callcaching/EngineJobHashingActor.scala]: Is there a race condition where the `callCacheReadingJobActor` shuts down while this `NextHit` message is being processed?

File: engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/callcaching/EngineJobHashingActor.scala [hunk 16]
```diff
@@ -1,282 +1,126 @@
...
+        case None =>
+          jobLogger.error("Requested next hit but there's no Call Caching Read Actor !")
```
Comment (cjllanwarne) [inline engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/callcaching/EngineJobHashingActor.scala]: Erroneous space before !

File: engine/src/test/scala/cromwell/engine/workflow/lifecycle/execution/callcaching/CallCacheHashingJobActorDataSpec.scala [hunk 17]
```diff
@@ -0,0 +1,94 @@
...
+import org.scalatest.prop.TableDrivenPropertyChecks
+import org.scalatest.{FlatSpecLike, Matchers}
+
+class CallCacheHashingJobActorDataSpec extends TestKitSuite with FlatSpecLike with BackendSpec with Matchers with Eventually with TableDrivenPropertyChecks {
+  behavior of "CallCacheReadingJobActorData"
+
+  val fileHash1 = HashResult(HashKey("key"), HashValue("value"))
+  val fileHash2 = HashResult(HashKey("key2"), HashValue("value2"))
+  val fileHash3 = HashResult(HashKey("key3"), HashValue("value3"))
+  val fileHashRequest1 = mock[SingleFileHashRequest]
```
Comment (cjllanwarne) [inline engine/src/test/scala/cromwell/engine/workflow/lifecycle/execution/callcaching/CallCacheHashingJobActorDataSpec.scala]: `SingleFileHashRequest` should be a `final case class` and so unmockable. 

That's ok, because you don't need to mock a case class, since this is a test we can just use `SingleFileHashRequest(null, hashKey, null null)` and allow the NPE if we set up the test wrong.

File: engine/src/test/scala/cromwell/engine/workflow/lifecycle/execution/callcaching/CallCacheHashingJobActorSpec.scala [hunk 18]
```diff
@@ -0,0 +1,292 @@
...
+  it should "send a correct InitialHashingResult upon starting" in {
+    val parent = TestProbe()
+    val inputs = Map(
+      "stringInput" -> WdlString("hello"),
+      "fileInput" -> WdlFile("world")
+    )
+    // Do not include "memory" on purpose, even though it's in the map of runtime attributes.
+    // This way we can verify that only attributes with a RuntimeAttributeDefinition are used for hashing
+    // Vice versa include a "docker" definition even though it's not in the map.
+    // This ensures that we still record the fact that there was no docker attribute with a "N/A" value
```
Comment (cjllanwarne) [inline engine/src/test/scala/cromwell/engine/workflow/lifecycle/execution/callcaching/CallCacheHashingJobActorSpec.scala]: Nice!

File: engine/src/test/scala/cromwell/engine/workflow/lifecycle/execution/callcaching/CallCacheHashingJobActorSpec.scala [hunk 19]
```diff
@@ -0,0 +1,292 @@
...
+  it should "keep going if the read actor dies and writeToCache is on" in {
+    val callCacheReadProbe = TestProbe()
+    val parent = TestProbe()
+    val cchja = makeCCHJA(Option(callCacheReadProbe.ref), TestProbe().ref, parent.ref)
+    parent.expectMsgClass(classOf[InitialHashingResult])
+    
+    cchja.setState(WaitingForHashFileRequest)
+    system stop callCacheReadProbe.ref
+
+    parent.expectMsg(NoFileHashesResult)
```
Comment (cjllanwarne) [inline engine/src/test/scala/cromwell/engine/workflow/lifecycle/execution/callcaching/CallCacheHashingJobActorSpec.scala]: It'd be nice if we could have some file hashes here (to make sure it did indeed keep going)

File: engine/src/test/scala/cromwell/engine/workflow/lifecycle/execution/callcaching/EJHADataSpec.scala [hunk 20]
```diff
@@ -1,96 +0,0 @@
...
-
-class EJHADataSpec extends FlatSpec with Matchers {
-
-  behavior of "EJHA data"
-
-  val hashKey1 = HashKey("Peter Piper picked a peck of pickled peppers")
-  val hashKey2 = HashKey("I saw Susie sitting in a shoe shine shop: Where she sits she shines, and where she shines she sits")
-  val hashKey3 = HashKey("How many boards could the Mongols hoard if the Mongol hordes got bored?")
-  val hashKey4 = HashKey("The sixth sick Sikh's sixth sheep is sick.")
-  val hashKey5 = HashKey("The Doge did what a Doge does, when a Doge does his duty to a Duke, that is. When the Doge did his duty and the Duke didn't, that's when the Duchess did the dirt to the Duke with the Doge. There they were in the dark: The Duke with his dagger, the Doge with his dart and the Duchess with her dirk. The Duchess dug at the Duke just when the Duke dove at the Doge. Now the Duke ducked, the Doge dodged, and the Duchess didn't. So the Duke got the Duchess, the Duchess got the Doge, and the Doge got the Duke.")
```
Comment (cjllanwarne) [inline engine/src/test/scala/cromwell/engine/workflow/lifecycle/execution/callcaching/EJHADataSpec.scala]: it's a sad day when this text is removed from the code base :'(

Comment (cjllanwarne) [inline engine/src/test/scala/cromwell/engine/workflow/lifecycle/execution/callcaching/EJHADataSpec.scala]: 😢

File: engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/ExecutionStore.scala (hunk 9 above)
Comment (Horneth) [inline engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/ExecutionStore.scala]: because `readyToStart.size` would force the stream (as in scala collection stream) to evaluate entirely and cause `arePrerequisitesDone` to be called on all subsequent scopes even though the will be ignored since we only take the first `ExecutionStore.MaxJobsToStartPerTick` ones.

File: database/migration/src/main/resources/changesets/call_caching_aggregation/call_caching_base_aggregation.sql [hunk 21]
```diff
@@ -0,0 +1,18 @@
+INSERT INTO CALL_CACHING_AGGREGATION_ENTRY(
+  CALL_CACHING_ENTRY_ID,
+  BASE_AGGREGATION
+)
+  SELECT cche.CALL_CACHING_ENTRY_ID,
+  UPPER(MD5(GROUP_CONCAT(
+                cche.HASH_VALUE ORDER BY cche.HASH_KEY ASC SEPARATOR ''
```
Comment (Horneth) [inline database/migration/src/main/resources/changesets/call_caching_aggregation/call_caching_base_aggregation.sql]: Why ?

File: engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/callcaching/CallCacheHashingJobActor.scala (hunk 14 above)
Comment (Horneth) [inline engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/callcaching/CallCacheHashingJobActor.scala]: Yes the migration query leaves out the runtime attributes that have `checkForHitOrMiss` false

File: database/sql/src/main/scala/cromwell/database/slick/CallCachingSlickDatabase.scala [hunk 22]
```diff
@@ -19,8 +19,8 @@ trait CallCachingSlickDatabase extends CallCachingSqlDatabase {
                              (implicit ec: ExecutionContext): Future[Unit] = {
 
     // Construct parallel lists of parent entries, hashes, simpletons, and detritus from `CallCachingJoin`s.
-    val (entries, hashes, simpletons, detritus) = joins.toList.foldMap { j =>
-      (List(j.callCachingEntry), List(j.callCachingHashEntries), List(j.callCachingSimpletonEntries), List(j.callCachingDetritusEntries)) }
+    val (entries, hashes, simpletons, detritus, aggregations) = joins.toList.foldMap { j =>
+      (List(j.callCachingEntry), List(j.callCachingHashEntries), List(j.callCachingSimpletonEntries), List(j.callCachingDetritusEntries), List(j.callCachingAggregationEntry.toList)) }
```
Comment (danbills) [inline database/sql/src/main/scala/cromwell/database/slick/CallCachingSlickDatabase.scala]: Just checking:  Is the intention to create a List[List[]] of Entries?

Comment (mcovarr) [inline database/sql/src/main/scala/cromwell/database/slick/CallCachingSlickDatabase.scala]: (This change was part of my rebase)

It was the intent that `entries` would be a `List[CallCachingEntry]` and all the others would be `List[Seq[_]]` or `List[List[_]]` of parallel dependent entities.

File: engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/callcaching/CallCache.scala (hunk 10 above)
Comment (Horneth) [inline engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/callcaching/CallCache.scala]: Oh right yes ! I guess technically this would never need to be called with an empty set but still..

File: engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/callcaching/CallCacheHashingJobActor.scala (hunk 12 above)
Comment (Horneth) [inline engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/callcaching/CallCacheHashingJobActor.scala]: Yeah I always wondered the reason for that. Akka already logs unhandled messages by default, is there a reason we want to do this ourselves instead ?

File: engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/callcaching/CallCacheHashingJobActor.scala (hunk 11 above)
Comment (Horneth) [inline engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/callcaching/CallCacheHashingJobActor.scala]: I believe so but I'll check

File: database/migration/src/main/resources/changesets/call_caching_aggregation/call_caching_base_aggregation.sql (hunk 21 above)
Comment (cjllanwarne) [inline database/migration/src/main/resources/changesets/call_caching_aggregation/call_caching_base_aggregation.sql]: Imagine I have two `Array[File]` inputs, `a` and `b`. Without keys I think you would cache match the following two input sets:
```
a = [ file_1, file_2 ]
b = [ file_3, file_4 ]
```
and:
```
a = [ file_1 ]
b = [ file_2, file_3, file_4 ]
```

Comment (cjllanwarne) [inline database/migration/src/main/resources/changesets/call_caching_aggregation/call_caching_base_aggregation.sql]: Or for the initial hash set, you could sub out `File`s for `String`s

File: engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/callcaching/CallCacheHashingJobActor.scala (hunk 12 above)
Comment (geoffjentry) [inline engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/callcaching/CallCacheHashingJobActor.scala]: FWIW I've never really liked our penchant for specially handling the unhandled case unless we were going to do something other than log. As you said, akka already does this.

File: engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/callcaching/EngineJobHashingActor.scala (hunk 15 above)
Comment (Horneth) [inline engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/callcaching/EngineJobHashingActor.scala]: I don't think so ? The `callCacheReadingJobActor` only shuts down if it determines it's a `CacheMiss` at which point the EJEA will stop asking for `NextHit`

**Comment by ruchim** [comment]
Perhaps I missed it, but it seems like we are still publishing the cache hit results as soon as we get a hit, not after copying succeeds? Maybe it wasn't meant to be a part of this PR.

**Comment by cjllanwarne** [comment]
@ruchim writing a cache result (its hashes and its results) should be functionally atomic (i.e. either you get a cache hit, or you don't, and if you do, then you can copy all the results immediately).
If that's not part of this PR then it should 100% be part of Cromwell 27 (IMO)

File: engine/src/main/scala/cromwell/engine/io/nio/NioFlow.scala [hunk 23]
```diff
@@ -77,4 +80,9 @@ class NioFlow(parallelism: Int, scheduler: Scheduler, nbAttempts: Int = IoActor.
       )
     }
   }
+  
+  private def createDirectoriesForSFSPath(path: Path) = path match {
```
Comment (ruchim) [inline engine/src/main/scala/cromwell/engine/io/nio/NioFlow.scala]: Is there risk in moving this step directly to the copyTo/write methods?

File: engine/src/test/scala/cromwell/engine/workflow/lifecycle/execution/callcaching/EJHADataSpec.scala (hunk 20 above)
Comment (ruchim) [inline engine/src/test/scala/cromwell/engine/workflow/lifecycle/execution/callcaching/EJHADataSpec.scala]: I'm sure you'll bring it back one day Chris :)

**Comment by cjllanwarne** [comment]
👍
